src/analyzing/CFG_build.ml

```diff
@@ -240,41 +259,122 @@ let rec cfg_stmt : state -> F.nodei option -> stmt -> cfg_stmt_result =
+      (* When this is true, control must go to the else clause. *)
+      let try_must_not_exit =
+        match finaltry with
+        | None -> false
+        | _ -> (not !(state.may_throw)) && not !(state.may_return)
+      in
+
+      (* From end of try to else. *)
+      state.g |> add_arc_from_opt (finaltry, elsei);
+
+      (* In the else clause, if an exception is thrown, go to the finally
+       * clause, or if there is no finally clause, propagate the exception.
+       *)
+      let throw_destination =
+        match finally_st with
+        | [] -> state.throw_destination
+        | _ -> Some newfakefinally
+      in
+      let else_state = { state with throw_destination; return_destination } in
+      let finalelse = cfg_stmt_list else_state (Some elsei) else_st in
+
+      (* If the else clause is reached, and this is true, then the flow stops
+       * after the finally clause.
+       *)
+      let else_must_exit =
+        match finalelse with
+        | None -> true
+        | _ -> false
+      in
+
+      state.g |> add_arc_from_opt (finalelse, newfakefinally);
+
+      (* From catches to finally in case of uncaught exceptions. *)
       state.g |> add_arc (catchesi, newfakefinally);
+
       catches
       |> List.iter (fun (_, catch_st) ->
-             let finalcatch = cfg_stmt_list state (Some catchesi) catch_st in
+             let catch_state =
+               { state with throw_destination; return_destination }
+             in
+             let finalcatch =
+               cfg_stmt_list catch_state (Some catchesi) catch_st
+             in
              state.g |> add_arc_from_opt (finalcatch, newfakefinally));
-      let finalfinally = cfg_stmt_list state (Some newfakefinally) finally_st in
-      (* If we're inside another Try then we assume that we could propagate up
-       * some unhandled exception. Otherwise we assume that we handled everything.
-       * THINK: Alternatively, we could add an arc to the exit node. *)
-      state.g |> add_arc_opt_to_opt (finalfinally, state.try_catches_opt);
-      CfgFirstLast (newi, finalfinally)
+
+      (* Inside the finally clause, throws and returns get propagated. *)
+      let finally_state = state in
+      let finalfinally =
+        cfg_stmt_list finally_state (Some newfakefinally) finally_st
+      in
+
+      (* Also propagate the exit (either from throws or returns) at the end
+       * of the finally clause. If there's no finally clause, the arcs
+       * should already be propagated from the logic above.
+       *)
+      if !(state.may_throw) then
+        state.g |> add_arc_opt_to_opt (finalfinally, state.throw_destination);
+      if !(state.may_return) then
+        state.g |> add_arc_from_opt (finalfinally, state.return_destination);
+
+      (* If we absolutely know that we need to propagate the exit, then
+       * we don't continue the dataflow after this statement.
+       *)
+      if try_must_return || (try_must_not_exit && else_must_exit) then
```

Comment:
Also, what if `try_must_return` is `false` but the `try` ends in `goto`. Does this work correctly?

Reply:
Will remove this logic to simplify this PR, but basically, it was needed above because it may be overridden afterwards. Eww, imperative programming :)